Signal/src/ViewControllers/ConversationView/ConversationInputToolbar.m

```diff
@@ -645,6 +688,99 @@ - (void)observeValueForKeyPath:(nullable NSString *)keyPath
     }
 }
 
+#pragma mark - Attachment Approval
+
+- (void)showApprovalUIForAttachment:(SignalAttachment *)attachment
+{
+    OWSAssert(attachment);
+
+    self.attachmentToApprove = attachment;
+
+    MediaMessageView *attachmentView = [[MediaMessageView alloc] initWithAttachment:attachment];
+    self.attachmentView = attachmentView;
+    [self.attachmentApprovalView addSubview:attachmentView];
+    [attachmentView autoPinEdgeToSuperviewEdge:ALEdgeTop withInset:10];
+    [attachmentView autoPinWidthToSuperviewWithMargin:20];
+
+    UIView *buttonRow = [UIView containerView];
+    [self.attachmentApprovalView addSubview:buttonRow];
+    [buttonRow autoPinWidthToSuperviewWithMargin:20];
+    [buttonRow autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:attachmentView withOffset:10];
+    [buttonRow autoPinEdgeToSuperviewEdge:ALEdgeBottom withInset:10];
+
+    // We use this invisible subview to ensure that the buttons are centered
+    // horizontally.
+    UIView *buttonSpacer = [UIView new];
+    [buttonRow addSubview:buttonSpacer];
+    // Vertical positioning of this view doesn't matter.
+    [buttonSpacer autoPinEdgeToSuperviewEdge:ALEdgeTop];
+    [buttonSpacer autoSetDimension:ALDimensionWidth toSize:ScaleFromIPhone5To7Plus(20, 30)];
+    [buttonSpacer autoSetDimension:ALDimensionHeight toSize:0];
+    [buttonSpacer autoHCenterInSuperview];
+
+    UIView *cancelButton = [self createAttachmentApprovalButton:[CommonStrings cancelButton]
+                                                          color:[UIColor ows_destructiveRedColor]
+                                                       selector:@selector(attachmentApprovalCancelPressed)];
+    [buttonRow addSubview:cancelButton];
+    [cancelButton autoPinHeightToSuperview];
+    [cancelButton autoPinEdge:ALEdgeRight toEdge:ALEdgeLeft ofView:buttonSpacer];
+
+    UIView *sendButton =
+        [self createAttachmentApprovalButton:NSLocalizedString(
+                                                 @"ATTACHMENT_APPROVAL_SEND_BUTTON", comment
+                                                 : @"Label for 'send' button in the 'attachment approval' dialog.")
+                                       color:[UIColor colorWithRGBHex:0x2ecc71]
+                                    selector:@selector(attachmentApprovalSendPressed)];
+    [buttonRow addSubview:sendButton];
+    [sendButton autoPinHeightToSuperview];
+    [sendButton autoPinEdge:ALEdgeLeft toEdge:ALEdgeRight ofView:buttonSpacer];
```

Comment:
Are we missing a `[sendButton autoPinEdge:ALEdgeRight toEdge:??? ofView:???];`?

Reply:
With this design, cancel and send buttons are fully constrained, but their size is set in `createAttachmentApprovalButton`.

Reply:
But I expect this design to change quite a bit.